pkg/providers/nutanix/workers.go

```diff
@@ -0,0 +1,136 @@
+	"github.com/aws/eks-anywhere/pkg/cluster"
+	"github.com/aws/eks-anywhere/pkg/clusterapi"
+	capiyaml "github.com/aws/eks-anywhere/pkg/clusterapi/yaml"
+	"github.com/aws/eks-anywhere/pkg/providers/common"
+	"github.com/aws/eks-anywhere/pkg/yamlutil"
+)
+
+type (
+	// Workers represents the nutanix specific CAPI spec for worker nodes.
+	Workers        = clusterapi.Workers[*v1beta1.NutanixMachineTemplate]
+	workersBuilder = capiyaml.WorkersBuilder[*v1beta1.NutanixMachineTemplate]
+)
+
+// WorkersSpec generates a nutanix specific CAPI spec for an eks-a cluster worker nodes.
+// It talks to the cluster with a client to detect changes in immutable objects and generates new
+// names for them.
+func WorkersSpec(ctx context.Context, logger logr.Logger, client kubernetes.Client, spec *cluster.Spec) (*Workers, error) {
+	ndcs := spec.NutanixDatacenter.Spec
+	machineConfigs := spec.NutanixMachineConfigs
+	controlPlaneMachineSpec, etcdMachineSpec := getControlPlaneAndEtcdMachineSpecs(spec, machineConfigs)
+
+	wnmcs := make(map[string]v1alpha1.NutanixMachineConfigSpec, len(spec.Cluster.Spec.WorkerNodeGroupConfigurations))
+	for _, machineConfig := range machineConfigs {
+		machineConfig.SetDefaults()
+	}
+
+	creds := GetCredsFromEnv()
+	templateBuilder := NewNutanixTemplateBuilder(&ndcs, controlPlaneMachineSpec, etcdMachineSpec, wnmcs, creds, time.Now)
+
+	workloadTemplateNames, kubeadmconfigTemplateNames := getTemplateNames(spec, templateBuilder, machineConfigs)
+
+	workersYaml, err := templateBuilder.GenerateCAPISpecWorkers(spec, workloadTemplateNames, kubeadmconfigTemplateNames)
+	if err != nil {
+		return nil, err
+	}
+
+	workers, err := parseWorkersYaml(logger, workersYaml)
+	if err != nil {
+		return nil, fmt.Errorf("parsing nutanix CAPI workers yaml: %w", err)
+	}
+
+	if err = workers.UpdateImmutableObjectNames(ctx, client, getMachineTemplate, machineTemplateEquals); err != nil {
+		return nil, fmt.Errorf("updating nutanix worker immutable object names: %w", err)
+	}
+
+	return workers, nil
+}
+
+func getControlPlaneAndEtcdMachineSpecs(spec *cluster.Spec, machineConfigs map[string]*v1alpha1.NutanixMachineConfig) (*v1alpha1.NutanixMachineConfigSpec, *v1alpha1.NutanixMachineConfigSpec) {
+	var controlPlaneMachineSpec, etcdMachineSpec *v1alpha1.NutanixMachineConfigSpec
+
+	if spec.Cluster.Spec.ControlPlaneConfiguration.MachineGroupRef != nil && machineConfigs[spec.Cluster.Spec.ControlPlaneConfiguration.MachineGroupRef.Name] != nil {
+		controlPlaneMachineSpec = &machineConfigs[spec.Cluster.Spec.ControlPlaneConfiguration.MachineGroupRef.Name].Spec
+	}
+
+	if spec.Cluster.Spec.ExternalEtcdConfiguration != nil {
+		if spec.Cluster.Spec.ExternalEtcdConfiguration.MachineGroupRef != nil && machineConfigs[spec.Cluster.Spec.ExternalEtcdConfiguration.MachineGroupRef.Name] != nil {
+			etcdMachineSpec = &machineConfigs[spec.Cluster.Spec.ExternalEtcdConfiguration.MachineGroupRef.Name].Spec
+		}
+	}
+
+	return controlPlaneMachineSpec, etcdMachineSpec
+}
```

Comment:
Why do we need to get control plane and etcd machine specs to generate worker spec?
Also, this function looks like a duplicate of https://github.com/aws/eks-anywhere/pull/5190/files#diff-3fb684191fa55c0f1a7409b4dd2c10f6b3faf13ac8e0ef441eece82b3600c8f5R83-R94

Reply:
We don't. I think I did workers and controlplane on two separate days and forgot about the overlap. Workers don't need controplane and etcd machinespecs and controlplane doesn't need workermachineconfigs. Good catch; fixed.